### PR TITLE
tests(e2e-dev): handle edge cases introduced with 404 for not existing pages

### PR DESCRIPTION
**Pull request by pieh**
## Description

https://github.com/gatsbyjs/gatsby/pull/33258 introduced some changes that need fixing/adjusting e2e dev runtime tests

### CODE REVIEW

File: e2e-tests/development-runtime/cypress/integration/functionality/query-data-caches.js [hunk 1]
```diff
@@ -110,7 +110,9 @@ function pageTitleAndDataAssertion(config) {
 function runTests(config) {
   preTestSetup(config)
 
-  cy.visit(`/query-data-caches/${config.slug}/page-A/`).waitForRouteChange()
+  cy.visit(`/query-data-caches/${config.slug}/page-A/`, {
+    failOnStatusCode: config.initialPage !== `404`,
+  }).waitForRouteChange()
```
Comment (pieh) [inline e2e-tests/development-runtime/cypress/integration/functionality/query-data-caches.js:115]: `cy.visit` would fail if status is 404 - we do want to continue (as we do expect this code for 404 page)

File: packages/gatsby/src/utils/start-server.ts [hunk 2]
```diff
@@ -705,7 +705,7 @@ module.exports = {
           res.send(e).status(500)
         }
       } else {
-        const potentialPagePath = reverseFixedPagePath(req.path)
+        const potentialPagePath = reverseFixedPagePath(decodeURI(req.path))
```
Comment (pieh) [inline packages/gatsby/src/utils/start-server.ts:708]: not fully sure about this one, but this particular one was causing issues with https://github.com/gatsbyjs/gatsby/blob/67c82e2b862fe39ca6af6d4060ace43c55f7f20d/e2e-tests/development-runtime/cypress/integration/navigation/linking.js#L113 (browsers under the hood will encode those).

Other option is to try `req.path` as is and decoded one separately? but given that this seems to make tests happy in general it should be fine?

**Review by wardpeet — APPROVED** [review]

File: packages/gatsby/src/utils/start-server.ts (hunk 2 above)
Comment (wardpeet) [reply packages/gatsby/src/utils/start-server.ts:708]: I wonder if this is something cloud needs to fix too
